### PR TITLE
Major refactoring of Code around the Record class

### PR DESCRIPTION
**Pull request by sarabrajsingh**
take a look at the program and tell me what you think

### CODE REVIEW

File: assignment+4/Record.java [hunk 1]
```diff
@@ -1,60 +1,99 @@
...
+public class Record implements Comparable<Record> {
+  private String cgndbId;
+  private String geographicName;
+  private String genericTerm;
+  private Double latitude;
+  private Double longitude;
+  private String location;
+  private String province;
+
+  public static String SORT_BY = "CGNDBID";
```
Comment (dunyaoguz) [inline assignment+4/Record.java:12]: What does this do?

File: assignment+4/Record.java [hunk 2]
```diff
@@ -1,60 +1,99 @@
...
-    this.latitude = Float.parseFloat(elements[3]);
-    this.longitude = Float.parseFloat(elements[4]);
-    this.location = null;
-
-    // if there is a location, assign it to the class variable
-    if (size == 6) {
-      this.location = elements[5];
-    }
+
+public class Record implements Comparable<Record> {
```
Comment (dunyaoguz) [inline assignment+4/Record.java:3]: What does Comparable<Record> do?

File: assignment+4/Main.java [hunk 3]
```diff
@@ -5,152 +5,168 @@
...
 public class Main {
 
-    // returns a BufferedReader object for a given file
-    public static BufferedReader fileReader(String filename) throws IOException {
-      File file = new File(filename);
-      FileReader fr = new FileReader(file);
-      BufferedReader br = new BufferedReader(fr);
-      return br;
-    }
+  public static int NUM_RECORDS = 10000; // globally available # of records count
```
Comment (dunyaoguz) [inline assignment+4/Main.java:17]: Why do we set this?

File: assignment+4/Main.java [hunk 4]
```diff
@@ -5,152 +5,168 @@
...
+        // build a record object based on CSV input
+        String cgndbID = null;
+        String geographicName = null;
+        String geographicTerm = null;
+        Double latitude = null;
+        Double longitude = null;
+        String location = null;
+        String province = null;
+
+        for(int i = 0; i < input.length; i++){
```
Comment (dunyaoguz) [inline assignment+4/Main.java]: what is this doing..

File: assignment+4/Main.java [hunk 5]
```diff
@@ -5,152 +5,168 @@
...
+          try{
+            switch(i){
+              case 0:
```
Comment (dunyaoguz) [inline assignment+4/Main.java:45]: Can you update the part of the readme that describes how the csv needs to look like ?

**Comment by dunyaoguz** [comment]
log outputs: 
+ title capitalized
+ all capitalized header like there exists for the previous outputs (something like "SEARCH RESULTS" with ==== underneath)
+ ">>>" at the beginning
+ grammar mistake with the word "generic" 

@sarabrajsingh

File: assignment+4/Main.java (hunk 4 above)
Comment (sarabrajsingh) [reply assignment+4/Main.java]: reading each line in any given CSV and trying to parse the data, surrounded by a try/catch block for fault tolerance (so our program doesn't shit the bed when a null field is found)

File: assignment+4/Record.java (hunk 2 above)
Comment (sarabrajsingh) [reply assignment+4/Record.java:3]: If we want to override and introduce custom behaviour to the Record class when we do something like:

`Record record1 = new Record(someID1);
Record record2 = new Record(someID2);
int = record1.compareTo(record2);`

in java, we have to implement the Comparable Interface. line 3 in plain english is saying -> I want a publically available class called "Record" which implements the Comparable interface of type Record.

File: assignment+4/Record.java (hunk 1 above)
Comment (sarabrajsingh) [reply assignment+4/Record.java:12]: that public static variable (accessible anywhere in the program) will be used a switch to run comparisons in the Record class, in the compareTo() method in the Record class

File: assignment+4/InvertedIndex.java [hunk 6]
```diff
@@ -66,100 +55,97 @@ private int findEntry(String term) {
...
+  public String toString(){
+    StringBuilder sb = new StringBuilder();
+    for (Entry e : entries){
+      sb.append(e.term + " : " + e.documents.toString())
+        .append("\n");
     }
+    return sb.toString().trim();
   }
 
   public static void main(String[] args) {
```
Comment (dunyaoguz) [inline assignment+4/InvertedIndex.java:127]: Do we need to comment out all the mains in all the objects?

File: assignment+4/InvertedIndex.java [hunk 7]
```diff
@@ -1,50 +1,40 @@
...
     }
-
-    private void printEntry() {
-      System.out.printf("Term: %s", term);
-      System.out.println();
-      System.out.println("Documents:");
-      for (Record d : documents) {
-        d.printRecord();
-      }
+    @Override
```
Comment (dunyaoguz) [inline assignment+4/InvertedIndex.java:21]: Is this necessary put whenever you do a toString() function?

File: assignment+4/InvertedIndex.java (hunk 6 above)
Comment (sarabrajsingh) [reply assignment+4/InvertedIndex.java:127]: teeeeechnically, there should only ever be one public static void main(String[] args) in our entire program hence why i commented out the other main methods not in Main.java

File: assignment+4/InvertedIndex.java [hunk 8]
```diff
@@ -1,50 +1,40 @@
...
+    public String toString(){
+      return new String(term + documents.toString());
     }
   }
 
   // constructor for the InvertedIndex object
   public InvertedIndex() {
-    this.entries = new Entry[0];
+    this.entries = new ArrayList<Entry>(Main.NUM_RECORDS);
```
Comment (dunyaoguz) [inline assignment+4/InvertedIndex.java]: Why set to num_records size, we won't necessarily have the same number of terms as number of records (since some terms may be duplicated)

File: assignment+4/InvertedIndex.java (hunk 7 above)
Comment (sarabrajsingh) [reply assignment+4/InvertedIndex.java:21]: every object in java inherits the toString() function from the Object superclass (parent of all objects in Java).

if you want to override what happens when a call on your object, like `System.out.println(object)` behaves, then you need to override the toString() method.

i thought it would be cleaner to have a line of code like:
`System.out.println(record)`

than to have something like

`System.out.println(record.cgndbid + record.latitude)` etc etc

File: assignment+4/InvertedIndex.java [hunk 9]
```diff
@@ -66,100 +55,97 @@ private int findEntry(String term) {
...
+      char c = string.charAt(i);
+      if (c <= '\u007F') {
+        output[counter++] = c;
+      }
+    }
+    String flattenedTerm = (new String(output)).toLowerCase();
+    // step 2 - split the new flattenedTerm, where delimiter = " ", and find the
+    // difference between this new list and the stopWords[] list. The difference
+    // will be our term without stopWords
+    StringBuilder sb = new StringBuilder();
```
Comment (dunyaoguz) [inline assignment+4/InvertedIndex.java:77]: what's this class

File: assignment+4/Main.java (hunk 5 above)
Comment (sarabrajsingh) [reply assignment+4/Main.java:45]: yes of course. its going to be just the unadultered CSVs from open.canada.ca

File: assignment+4/InvertedIndex.java (hunk 6 above)
Comment (dunyaoguz) [inline assignment+4/InvertedIndex.java:127]: @sarabrajsingh you didn't comment out main in InvertedIndex, Record and Query classes

**Review by dunyaoguz — APPROVED** [review]
